resource-management/pkg/aggregrator/list-watch-impl.go

```diff
@@ -0,0 +1,167 @@
+	common_lib "global-resource-service/resource-management/pkg/common-lib"
+	"global-resource-service/resource-management/pkg/common-lib/metrics"
+	"global-resource-service/resource-management/pkg/common-lib/types"
+	event "global-resource-service/resource-management/pkg/common-lib/types/runtime"
+)
+
+// LwRun implements Run interface of Aggregator
+func (a *Aggregator) Run() (err error) {
+	numberOfRegions := len(a.urls)
+
+	klog.V(3).Infof("Running for loop to connect to to resource region manager...")
+
+	for i := 0; i < numberOfRegions; i++ {
+		go func(i int) {
+			klog.V(3).Infof("Starting goroutine for region: %v", a.urls[i])
+			defer func() {
+				klog.V(3).Infof("Existing goroutine for region: %v", a.urls[i])
+			}()
+
+			var crv types.TransitResourceVersionMap
+			var regionNodeEvents [][]*event.NodeEvent
+			var length uint64
+			var eventProcess bool
+
+			// create client to resource region manager
+			c := NewRrmsClient(Config{ServiceUrl: a.urls[i], RequestTimeout: 30 * time.Minute})
+
+			klog.V(3).Infof("Starting loop list-watching nodes from region: %v", a.urls[i])
+
+			// TODO: add limit in config
+			// TODO: add pagination feature in region resource mgr service
+			// hack, list all 1m node in one call without pagination
+			regionNodeEvents, crv, length = a.listNodes(c, ListOptions{Limit: 1000000})
+
+			if length != 0 {
+				klog.V(4).Infof("Total (%v) region node events are listed successfully in (%v) RPs", length, len(regionNodeEvents))
+			} else {
+				// TODO: handel empty list
+			}
+
+			// Convert 2D array to 1D array
+			// TODO: add dynamically watch at RP granularity level feature in GRS
+			//       by spawning threads for watch each RP, we can further optimize concurrency and also avoid below conversions
+			minRecordNodeEvents := make([]*event.NodeEvent, 0, length)
+			for j := 0; j < len(regionNodeEvents); j++ {
+				minRecordNodeEvents = append(minRecordNodeEvents, regionNodeEvents[j]...)
+			}
+
+			start := time.Now()
+			eventProcess, crv = a.EventProcessor.ProcessEvents(minRecordNodeEvents)
+			end := time.Now()
+			klog.V(6).Infof("Event Processor Processed nodes results : %v. duration: %v", eventProcess, end.Sub(start))
+
+			// start watch node changes
+			a.watchNodes(c, crv)
+
+		}(i)
+	}
+
+	klog.V(3).Infof("Finished for loop to connect to to resource region manager...")
+	return nil
+}
+
+func (a *Aggregator) listNodes(client RrmsInterface, listOpts ListOptions) (nodeList [][]*event.NodeEvent, crv types.TransitResourceVersionMap, length uint64) {
+	var start, end time.Time
+	var err error
+
+	for {
+		klog.Infof("List resources from region manager ...")
+		start = time.Now().UTC()
+		nodeList, crv, length, err = client.List(listOpts)
+		end = time.Now().UTC()
+		if err != nil {
+			klog.Errorf("failed list resource from region manager. error %v. retry in one second", err)
+			time.Sleep(1 * time.Second)
+			continue
```

Comment:
This is an infinite loop. Suggest set some limit here

Reply:
good point. will do.